Stops listening on UDP port after getting invalid json

This will cause prompush to return from the `listenUDP` function, causing it to close `serverConn`:

`echo "hello" | nc --udp localhost -p 9090`

logs:

`2016/11/09 17:20:37 udp.go:36: Error:  invalid character 'h' looking for beginning of value`

then a second `nc` prints:

`Ncat: Connection refused.`

--- a/.travis.yml
+++ b/.travis.yml
@@ -1,0 +1,10 @@
+language: go
+
+go:
+  - 1.4
+  - 1.5
+  - 1.6
+  - 1.7
+  - tip
+
+script: make test

--- a/Makefile
+++ b/Makefile
@@ -43,14 +43,6 @@ container: release_linux
 test:
 	@echo "* Running tests"
 	@$(GO) test $(PROJECT)/...
-	@echo
-
-	@echo "* Checking code with golint"
-	@golint src/$(PROJECT)/...
-	@echo
-
-	@echo "* Checking code with go vet"
-	@$(GO) vet $(PROJECT)/...
 
 clean:
 	rm -fr bin pkg vendor/pkg

--- a/README.md
+++ b/README.md
@@ -1,6 +1,8 @@
 
 # Pushprom
 
+[![Build Status](https://travis-ci.org/messagebird/pushprom.svg?branch=master)](https://travis-ci.org/messagebird/pushprom)
+
 Pushprom is a proxy (HTTP/UDP) to the [Prometheus](https://prometheus.io/) Go client.
 
 Prometheus doesn't offer a PHP client and PHP clients are hard to implement because they would need to keep track of state and PHP setups generally don't encourage that. That's why we built Pushprom.

--- a/src/pushprom/udp.go
+++ b/src/pushprom/udp.go
@@ -25,7 +25,7 @@ func listenUDP() {
 		n, _, err := serverConn.ReadFromUDP(buf)
 		if err != nil {
 			log.Println("Error: ", err)
-			return
+			continue
 		}
 		udpPacketCount.Inc()
 		if *debug {
@@ -34,13 +34,12 @@ func listenUDP() {
 		delta, err := NewDelta(bytes.NewBuffer(buf[0:n]))
 		if err != nil {
 			log.Println("Error: ", err)
-			return
+			continue
 		}
 
 		err = delta.Apply()
 		if err != nil {
 			log.Println("Error: ", err)
 		}
-
 	}
 }

--- a/src/pushprom/udp_test.go
+++ b/src/pushprom/udp_test.go
@@ -2,32 +2,46 @@ package main
 
 import (
 	"encoding/json"
+	"log"
 	"net"
+	"os"
 	"testing"
 	"time"
 
 	"github.com/stretchr/testify/assert"
 )
 
-func TestUDP(t *testing.T) {
+var conn *net.UDPConn
 
+func TestMain(m *testing.M) {
 	*udpListenAddress = ":3007"
+
 	go listenUDP()
 
 	// wait for it to "boot"
 	time.Sleep(time.Millisecond * 1000)
 
 	serverAddr, err := net.ResolveUDPAddr("udp", "127.0.0.1"+*udpListenAddress)
-	assert.Nil(t, err)
+	if err != nil {
+		log.Fatal("Could not resolve server UPD address")
+	}
 
 	localAddr, err := net.ResolveUDPAddr("udp", "127.0.0.1:0")
-	assert.Nil(t, err)
+	if err != nil {
+		log.Fatal("Could not resolve local UDP address")
+	}
 
-	conn, err := net.DialUDP("udp", localAddr, serverAddr)
-	assert.Nil(t, err)
+	conn, err = net.DialUDP("udp", localAddr, serverAddr)
+	if err != nil {
+		log.Fatal("Could not establish UDP connection")
+	}
 
 	defer conn.Close()
 
+	os.Exit(m.Run())
+}
+
+func TestUDP(t *testing.T) {
 	delta := &Delta{
 		Type:   GAUGE,
 		Method: "set",
@@ -38,7 +52,7 @@ func TestUDP(t *testing.T) {
 
 	buf, _ := json.Marshal(delta)
 
-	_, err = conn.Write(buf)
+	_, err := conn.Write(buf)
 	assert.Nil(t, err)
 
 	time.Sleep(time.Millisecond * 500)
@@ -47,5 +61,51 @@ func TestUDP(t *testing.T) {
 	result, err := readMetric(metrics, delta.Name, delta.Labels)
 	if assert.Nil(t, err) {
 		assert.Equal(t, delta.Value, result)
+	}
+}
+
+func TestIncorrectJson(t *testing.T) {
+	// First, let's write the correct value
+	oldDelta := &Delta{
+		Type:   GAUGE,
+		Method: "set",
+		Name:   "tree_width",
+		Help:   "the width in meters of the tree",
+		Value:  2.2,
+	}
+	buf, _ := json.Marshal(oldDelta)
+
+	_, err := conn.Write(buf)
+	assert.Nil(t, err)
+
+	oldMetrics := fetchMetrics(t)
+	oldResult, err := readMetric(oldMetrics, oldDelta.Name, oldDelta.Labels)
+	if assert.Nil(t, err) {
+		assert.Equal(t, oldDelta.Value, oldResult)
+	}
+
+	// Now, let's write the incorrect value
+	buf = []byte("hello")
+
+	_, err = conn.Write(buf)
+	assert.Nil(t, err)
+
+	// Last, let's write a new value
+	newDelta := &Delta{
+		Type:   GAUGE,
+		Method: "set",
+		Name:   "tree_width",
+		Help:   "the width in meters of the tree",
+		Value:  2.5,
+	}
+	buf, _ = json.Marshal(newDelta)
+
+	_, err = conn.Write(buf)
+	assert.Nil(t, err)
+
+	newMetrics := fetchMetrics(t)
+	newResult, err := readMetric(newMetrics, newDelta.Name, newDelta.Labels)
+	if assert.Nil(t, err) {
+		assert.Equal(t, newDelta.Value, newResult)
 	}
 }

